### ISSUE TITLE
Checksums for released archive files are not published to dotnetclichecksums

### ISSUE THREAD

**Issue by mthalman** [dotnet/aspnetcore#18792]
The expectation is that there should be checksum files located at the dotnetclichecksums blob storage location for ASP.NET archive files.  Example: https://dotnetclichecksums.blob.core.windows.net/dotnet/aspnetcore/Runtime/5.0.0-alpha.1.20071.6/aspnetcore-runtime-5.0.0-alpha.1.20071.6-linux-arm64.tar.gz.sha512 (doesn't exist but it should).  This is not being done for ASP.NET but is done for runtime and SDK (e.g. https://dotnetclichecksums.blob.core.windows.net/dotnet/Runtime/5.0.0-alpha.1.20103.2/dotnet-apphost-pack-5.0.0-alpha.1.20103.2-win-x86.msi.sha512).

This is needed for both 3.1 servicing and 5.0.

**Comment by dougbu** [comment]
/cc @Pilchie @anurse 

@mthalman and @MichaelSimons please list the exact list of "supported versions" that need these files. Also, when was the last time ASP.NET Core successfully pushed these checksums to the blob store?

**Comment by mthalman** [comment]
> @mthalman Matt Thalman FTE and @MichaelSimons Michael Simons (.NET) FTE please list the exact list of "supported versions" that need these files. Also, when was the last time ASP.NET Core successfully pushed these checksums to the blob store?

This should be done for all future releases of 2.1, 3.1, and 5.0 for all platforms.

I don't believe this was ever working for 3.0+.  For 2.1, it stopped working with the release of 2.1.2.  For 2.2, it stopped working with the release of 2.2.1.

**Comment by wtgodbe** [comment]
Looking into it, notes here.

- dotnet/runtime generates these files in their `Prepare artifacts and upload to build` step

**Comment by wtgodbe** [comment]
We used to use the Buildtools task, I guess: https://github.com/dotnet/buildtools/blob/6736870b84e06b75e7df32bb84d442db1b2afa10/src/Microsoft.DotNet.Build.Tasks/GenerateChecksums.cs. The other option, for now, is add a dependency on the buildtools package that carries this task, but that makes me nervous since buildtools is "deprecated"

**Comment by Pilchie** [comment]
The original mailthread here suggested that arcade had support for this directly.

**Comment by dagood** [comment]
My mistake--Arcade supports uploading to `dotnetclichecksums` but yes, doesn't appear to have the build task or anything like automatic processing during the publish stage. I think it would be reasonable to track getting it into arcade (or copying it to a local build tasks project if this repo has one). Definitely don't try to use buildtools. 🙂

**Comment by wtgodbe** [comment]
Core-SDK does this with a checked-in task also: https://github.com/dotnet/core-sdk/blob/7464d9bca4a2f25d1f832f9ef51ed288d4b0201e/src/core-sdk-tasks/GenerateChecksums.cs

**Comment by wtgodbe** [comment]
@chcosta @dagood @JohnTortugo @riarenas is there a "preferred process" for porting stuff like this to Arcade? My inclination would be that `GenerateChecksums` should go into https://github.com/dotnet/arcade/tree/master/src/Microsoft.DotNet.Build.Tasks.Packaging, but I don't know how fast-and-loose you guys play with this process

**Comment by dagood** [comment]
I'd argue for the SDK. https://github.com/dotnet/arcade/tree/master/src/Microsoft.DotNet.Arcade.Sdk/src. Packaging is for `pkgproj` (CoreFX-style packaging), doesn't make sense to bring it in for this. I'm not aware of any process particular to adding tasks to the Arcade SDK.

**Comment by wtgodbe** [comment]
> I'd argue for the SDK.

Glad I asked. Another question here - we need to fix this in 3.1 as well. Is it worth taking this in 3.x Arcade, or should we just check the task into AspNetCore's 3.1 branch, and only put it in Arcade for master/5.0? I'd lean towards the latter, I guess, since it isn't critical for servicing for repos other than AspNetCore. @mmitche does that sound right?

**Comment by wtgodbe** [comment]
@mthalman what exactly is the closure of asset types that need checksum files? Our shipping assets include `.jar`, `.pom`, and `.tgz` files for signalR, plus `.msi`, `.tar.gz`, `.rpm`, `.zip`, `.exe`, `.deb`, `.wixlib` files that correspond to the runtime/installers.

**Comment by MichaelSimons** [comment]
@wtgodbe, My suggestion would be to produce checksums for the same artifacts we do for official releases as [shown in the releases.json](https://github.com/dotnet/core/blob/master/release-notes/3.1/releases.json#L287)

**Comment by wtgodbe** [comment]
> IMO take it into 3.x arcade.

Would Core-Sdk/Core-Setup then change their builds to use the Arcade task, as a servicing fix? That seems unlikely. If we have another Arcade change in 3.x queued up, which will need to be ingested by all the 3.1 branches, then I think it'd be fine to piggyback off of that. But if not I'd be more inclined to reduce noise by only doing the Arcade port for master

**Comment by dagood** [comment]
> Would Core-Sdk/Core-Setup then change their builds to use the Arcade task, as a servicing fix? That seems unlikely.

I agree, since there's no direct reason to do so. However, if there becomes some reason to change the checksum code in 3.1, it would be good to have it centralized in Arcade so Core-Setup and Core-SDK can swap over and delete the duplicate rather than copy the C# fix. (Kinda weak justification though, I'll fully admit.)

**Comment by wtgodbe** [comment]
I'm coming around to the idea of porting this to Arcade for 3.x. I'll get cracking on that

**Comment by wtgodbe** [comment]
The task is now in 3.x arcade, and https://github.com/dotnet/arcade/pull/4500 merges it to master. That PR has been open for a while, so I may just cherry-pick it manually.

**Comment by wtgodbe** [comment]
@MichaelSimons I just took another look at https://github.com/dotnet/core/blob/master/release-notes/3.1/releases.json#L287, and I'm confused. That file lists assets like `aspnetcore-runtime-win-x64.zip`, but I don't see those assets getting published by AspNetCore - I only see stuff like `aspnetcore-runtime-3.1.1-win-x64.zip`. Where are those names coming from?

**Comment by MichaelSimons** [comment]
@wtgodbe, I believe the name is a type of friendly name.  The urls reference the correctly versioned components eg https://github.com/dotnet/core/blob/master/release-notes/3.1/releases.json#L287.  @leecow would be able to answer where the names come from.

**Comment by leecow** [comment]
Yes, that's a friendly name (the filename with the version stripped out) that enables easier querying for a given asset.

**Comment by wtgodbe** [comment]
Update: I have this ported to Arcade in 3.x & am getting it into master. Arcade's 3.x branch's official builds are blocked on getting a new DARC rollout, which will happen tomorrow morning. So hopefully tomorrow I'll have a PR up in AspNetCore to get the checksums generated.

**Comment by Pilchie** [comment]
@wtgodbe Should this be closed now?

**Comment by wtgodbe** [comment]
Not yet, I still need a solution for 2.1

**Comment by wtgodbe** [comment]
Looks like we got this wrong - the checksums for the zip files are incorrect. My current theory is that this is due to the zip files being built in steps prior to the step that does publishing - the .msi's, which are built in the step that does publishing (build-installers), are correct. We'll need to consider disabling this feature until we get it right (which means commenting out https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Faspnetcore%2Fblob%2Fce0579255104f6ea1902b62c1e21195a7b93f828%2Feng%2FPublishing.props%23L52-L55&data=02%7C01%7Cwigodbe%40microsoft.com%7C927e797afb5c4bccc31e08d7bc6f578e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637185058728704154&sdata=%2FWm17CV01beRo4qgCT7fc8wGf8UzuaIxzVTDhVY0YMs%3D&reserved=0)

**Comment by dougbu** [comment]
Instead, exclude the bad checksum files from the `@(_ChecksumsToPublish)` or fix the underlying problem.

**Comment by dougbu** [comment]
BTW copying is unlikely to mess up checksums. Are the Zip files signed or otherwise changed after the checksums are written?

**Comment by wtgodbe** [comment]
Ah, yup, that's it - they're signed after we generate the checksums: https://github.com/dotnet/aspnetcore/blob/ce0579255104f6ea1902b62c1e21195a7b93f828/eng/Signing.props#L16.

So we need to do a full one-off upload of all checksums after everything's been built. Don't have confidence I can make the necessary changes to the process in time to make 3.1 - I think it's best to punt the checksums to 3.1.4 entirely given that we know some of them are wrong, rather than uploading a subset of them (I haven't gone through and validated all of them individually)

**Comment by wtgodbe** [comment]
https://github.com/dotnet/aspnetcore/pull/19442

**Comment by mthalman** [comment]
@wtgodbe - Is this going to be integrated into 2.1 as well?

**Comment by wtgodbe** [comment]
Yes, I'm aiming to get this resolved for the next servicing release (3.1.4 in May)

**Comment by dougbu** [comment]
@wtgodbe can this issue now be closed? If not, what work remains?

**Comment by wtgodbe** [comment]
Not yet, still need to finish the 2.1 fix

**Comment by wtgodbe** [comment]
https://github.com/dotnet/versions/pull/580 should resolve this for release/2.1 (thanks @mmitche)

**Comment by wtgodbe** [comment]
Closing as fixes have been merged into all affected branches
